src/Libraries/Microsoft.Extensions.Caching.Hybrid/Internal/HybridCacheEventSource.cs

```diff
@@ -0,0 +1,192 @@
+        Volatile.Write(ref _totalBackendExecute, 0);
+        Volatile.Write(ref _currentBackendExecute, 0);
+        Volatile.Write(ref _currentDistributedFetch, 0);
+        Volatile.Write(ref _totalLocalCacheWrite, 0);
+        Volatile.Write(ref _totalDistributedCacheWrite, 0);
+        Volatile.Write(ref _totalStampedeJoin, 0);
+    }
+
+    [Event(EventIdLocalCacheHit, Level = EventLevel.Verbose)]
+    public void LocalCacheHit()
+    {
+        DebugAssertEnabled();
+        _ = Interlocked.Increment(ref _totalLocalCacheHit);
+        WriteEvent(EventIdLocalCacheHit);
+    }
+
+    [Event(EventIdLocalCacheMiss, Level = EventLevel.Verbose)]
+    public void LocalCacheMiss()
+    {
+        DebugAssertEnabled();
+        _ = Interlocked.Increment(ref _totalLocalCacheMiss);
+        WriteEvent(EventIdLocalCacheMiss);
+    }
+
+    [Event(EventIdDistributedCacheGet, Level = EventLevel.Verbose)]
+    public void DistributedCacheGet()
+    {
+        // should be followed by DistributedCacheHit, DistributedCacheMiss or DistributedCacheFailed
+        DebugAssertEnabled();
+        _ = Interlocked.Increment(ref _currentDistributedFetch);
+        WriteEvent(EventIdDistributedCacheGet);
+    }
+
+    [Event(EventIdDistributedCacheHit, Level = EventLevel.Verbose)]
+    public void DistributedCacheHit()
+    {
+        DebugAssertEnabled();
+        _ = Interlocked.Increment(ref _totalDistributedCacheHit);
+        _ = Interlocked.Decrement(ref _currentDistributedFetch);
+        WriteEvent(EventIdDistributedCacheHit);
+    }
+
+    [Event(EventIdDistributedCacheMiss, Level = EventLevel.Verbose)]
+    public void DistributedCacheMiss()
+    {
+        DebugAssertEnabled();
+        _ = Interlocked.Increment(ref _totalDistributedCacheMiss);
+        _ = Interlocked.Decrement(ref _currentDistributedFetch);
+        WriteEvent(EventIdDistributedCacheMiss);
+    }
+
+    [Event(EventIdDistributedCacheFailed, Level = EventLevel.Error)]
+    public void DistributedCacheFailed()
+    {
+        DebugAssertEnabled();
+        _ = Interlocked.Decrement(ref _currentDistributedFetch);
+        WriteEvent(EventIdDistributedCacheFailed);
+    }
+
+    [Event(EventIdBackendExecuteStart, Level = EventLevel.Verbose)]
+    public void BackendExecuteStart()
+    {
+        // should be followed by BackendExecuteComplete or BackendExecuteFailed
+        DebugAssertEnabled();
+        _ = Interlocked.Increment(ref _totalBackendExecute);
+        _ = Interlocked.Increment(ref _currentBackendExecute);
+        WriteEvent(EventIdBackendExecuteStart);
+    }
+
+    [Event(EventIdBackendExecuteComplete, Level = EventLevel.Verbose)]
+    public void BackendExecuteComplete()
+    {
+        DebugAssertEnabled();
+        _ = Interlocked.Decrement(ref _currentBackendExecute);
+        WriteEvent(EventIdBackendExecuteComplete);
+    }
+
+    [Event(EventIdBackendExecuteFailed, Level = EventLevel.Error)]
+    public void BackendExecuteFailed()
```

Comment:
BackendExecute doesn't sound obvious. I only know what it's referring to because I'm reading the code.

Reply:
the flags enum uses terminology "underlying data", so: switching to `UnderlyingDataQueryStart`, `UnderlyingDataQueryComplete`, `UnderlyingDataQueryFailed`, etc; I think that works better